assets/js/base/hocs/with-query-string-values.js

```diff
@@ -0,0 +1,92 @@
+import { Component } from 'react';
+import { addQueryArgs, getQueryArg } from '@wordpress/url';
+
+/**
+ * HOC that keeps the state in sync with the URL query string.
+ */
+const withQueryStringValues = ( values ) => ( OriginalComponent ) => {
+	let instances = 0;
+
+	class WrappedComponent extends Component {
+		urlParameterSuffix = instances++ > 0 ? `_${ instances }` : '';
+
+		hasWindowDependencies =
+			typeof window === 'object' &&
+			window.hasOwnProperty( 'history' ) &&
+			window.hasOwnProperty( 'location' ) &&
+			typeof window.addEventListener === 'function' &&
+			typeof window.removeEventListener === 'function';
```

Comment:
If we find this is a repeated pattern (might not be), we may want to consider extracting this to a utility function.  No need to do now though, just an observation.